Clarify dependent groups (#487)
Kattis/problem-package-format: spec/2023-07-draft.md

`max_score`               | Integer or `unbounded`              | 100 in `secret`, otherwise `unbounded`         | The maximum possible score of the test data group. Must be a non-negative integer or `unbounded`. This key is only permitted for the `secret` group and its subgroups.
`score_aggregation`       | `pass-fail`, `sum`, or `min`        | `sum` in `secret`, otherwise `pass-fail`       | How the score is determined based on the scores of the contained groups or test cases. [See Result Aggregation](#result-aggregation). This key is only permitted for the `secret` group and its subgroups.
`static_validation_score` | Integer or `pass-fail`            |                                                | The maximum score of the static validation test case, or `pass-fail`. See [Static Validator](#static-validator).

> `require_pass`            | String or sequence of strings       | empty sequence                                 | Names of other test data groups, or `sample`, whose test cases a submission must pass in order to receive a score for this test group. [See Result Aggregation](#result-aggregation). This key is only permitted for the `secret` group and its subgroups.

Reviewer: ...if specified in sample, it must be empty, since no group can come before sample lexicographically. I think it's better to disallow it.

> `require_pass`            | String or sequence of strings       | empty sequence                                 | Names of other test data groups, or `sample`, whose test cases a submission must pass in order to receive a score for this test group. [See Result Aggregation](#result-aggregation). This key is only permitted for the `secret` group and its subgroups.

Author: Yes, what @Matistjati said.

> `require_pass`            | String or sequence of strings       | empty sequence                                 | Names of other test data groups, or `sample`, whose test cases a submission must pass in order to receive a score for this test group. [See Result Aggregation](#result-aggregation). This key is only permitted for the `secret` group and its subgroups.

Reviewer 2: Okay I agree.

This is the new suggestion

```suggestion
`require_pass`            | String or sequence of strings       | empty sequence                                 | Names of other test data groups, or `sample`, whose test cases a submission must pass in order to receive a score for this test group. See [Group Dependencies](#group-dependencies). This key is only permitted for the `secret` group and its subgroups.
```

The secret data may be subdivided into test data groups.
Every subdirectory of `data/secret/` is a test data group and may contain a `test_group.yaml` configuration file.
`data/secret` can only have test data groups *or* test cases, never both.
The name of a test data group is the path under secret.

> For example, the test data group defined by the directory `data/secret/group1` has the name `secret/group1`.

Reviewer 2: Maybe move the lines
```
The name of a test data group is the path under secret.
For example, the test data group defined by the directory `data/secret/group1` has the name `secret/group1`.
```

under

```
That is, if there are any directories under `data/secret/` there must not be any `.in` files directly in `data/secret/` and vice versa.
```

Reviewer: I'm convinced. The sentence
```
That is, if there are any directories under `data/secret/` there must not be any `.in` files directly in `data/secret/` and vice versa.
```

is "obviously" a follow-up to

```
data/secret` can only have test data groups *or* test cases, never both.
```

But this pr would break these up. If you prefer, we can also move down the sentence 
```
data/secret` can only have test data groups *or* test cases, never both.
```

Author: Thanks for explaining. I agree.

> `data/secret` can only have test data groups *or* test cases, never both.
> The name of a test data group is the path under secret.
> For example, the test data group defined by the directory `data/secret/group1` has the name `secret/group1`.
> That is, if there are any directories under `data/secret/` there must not be any `.in` files directly in `data/secret/` and vice versa.

Reviewer: ```suggestion
`data/secret` can only have test data groups *or* test cases, never both.
That is, if there are any directories under `data/secret/` there must not be any `.in` files directly in `data/secret/` and vice versa.
The name of a test data group is the path under secret.
For example, the test data group defined by the directory `data/secret/group1` has the name `secret/group1`.
```

It is a judge error if the score of any group or subgroup exceeds its `max_score`.

#### Group Dependencies

> A test data group, or `secret` may specify that it depends on some other test data groups, or `sample` by setting `require_pass`.

Author: That is correct, and those are good points.

> A test data group, or `secret` may specify that it depends on some other test data groups, or `sample` by setting `require_pass`.

Reviewer: It also won't be a problem for tooling (testdata_tools): it's trivial to simply not generate the key for sample. And it results in a more readable `test_groups.yaml` file.

Each such test data group must have `pass-fail` aggregation.
The group specifying `require_pass` will only be run if the group being referred to receives an accepted verdict for all test cases in the group.

> If the group is not run, the group score is 0, and it did not receive an accepted verdict for all test cases in the group.

Author: Merged a fix on this from @Matistjati

> The `require_pass` is set to the name of a single test data group, or `sample`, or a sequence of such names, or `sample`.

Reviewer: To me, the first is more confusing. I prefer Harry's version.

> The `require_pass` is set to the name of a single test data group, or `sample`, or a sequence of such names, or `sample`.

Author: I definitely agree that @hairez's is better... it's just arguably technically not completely correct.

It is a judge error if the score of `secret` or any test data group exceeds its `max_score`.

#### Group Dependencies

> A test data group, or `secret` may specify that it depends on some other test data groups, or `sample` by setting `require_pass`.

Author: ```suggestion
A test data group, or `secret` may specify that it depends on some other test data groups. 
Note that `secret` may only depend on `sample`. 
Also, note that `sample` cannot depend on any group.
```

It is a judge error if the score of `secret` or any test data group exceeds its `max_score`.

#### Group Dependencies

A test data group, or `secret` may specify that it depends on some other test data groups. 
Note that `secret` may only depend on `sample`. 
Also, note that `sample` cannot depend on any group.
Each group mentioned under `require_pass` must have `pass-fail` aggregation.
The group specifying `require_pass` will only be run if the group being referred to receives an accepted verdict for all test cases in the group.

> If the group is not run, the group score is 0, and it did not receive an accepted verdict for all test cases in the group.

Reviewer: I just think this doesn't make grammatical sense.
```suggestion
If the group is not run, the group score is 0, and all test cases in the group are treated as having received a verdict that is not accepted.
```

Reviewer 2: I agree with you, this is much better and fixes https://github.com/Kattis/problem-package-format/pull/487#discussion_r2314978568
